### ISSUE TITLE
Unable to use ECS hosted Grafana behind AWS ALB

### ISSUE THREAD

**Issue by kylegordon** [grafana/grafana#12177]
I tried to post this in the community forums, but was denied as it has more than two links. 

Hi,

I'm trying to launch Grafana on AWS ECS (Docker underneath), and having some issues with what Grafana thinks it should be addressed by. When using ECS you have to use an AWS ALB in order to have a consistent front end. The ALB is, in some ways, like HAProxy, NGinx, etc. It does path pattern matching, and in my case I'm trying to present Grafana as a path within a larger application, ie, we have https://app.foo.bar/sales, https://app.foo.bar/marketing, and hopefully https://app.foo.bar/grafana - the ALB takes care of directing the relevant requests to the appropriate ECS tasks.

I have Grafana configured with what I believe to be appropriate values:

- Name: GF_SERVER_DOMAIN
  Value: "app.foo.bar"
- Name: GF_SERVER_ROOT_URL
  Value: "https://app.foo.bar/grafana"

On all occasions it returns the Grafana unhandled 404 page - the white page with '{{alert.title}} and 4 links at the top left. 

At startup it emits the following log:
t=2018-06-06T09:12:45+0000 lvl=info msg="Initializing HTTP Server" logger=http.server address=0.0.0.0:3000 protocol=http subUrl=/grafana socket=

When I visit https://app.foo.bar/grafana/, Chrome Dev tools suggest that it's trying to load https://app.foo.bar/grafana/public/build/grafana.dark.css?v5.1.3 but being met with a 404.

The logs from the docker container when trying to visit the main page are as follows:

t=2018-06-06T09:40:46+0000 lvl=info msg="Request Completed" logger=context userId=0 orgId=1 uname= method=GET path=/grafana/public/build/grafana.dark.css status=404 remote_addr=10.2.10.206 time_ms=1 size=14912 referer=https://app.foo.bar/grafana/

I'm not sure why it's failing to deliver content from /grafana/. Can anyone help? It appears to be a bug.

**Comment by torkelo** [comment]
if you host grafana under a subpath sub path your load balancer (or reverse proxy) needs to rewrite the urls so the request sent to grafana does not include the subpath

**Comment by kylegordon** [comment]
@jangaraj Thanks for the info. I gave that a try, and whilst it renders just great at /, it fails at the subpath with a Grafana served 404 (graph, 'sorry for the inconvenience', etc)

@torkelo As I mentioned above, this is not 'my' load balancer, it's an AWS load balancer. AWS Application Load Balancers do not rewrite paths. 

For others who come across this ticket - Grafana 5.x will not work on its own behind AWS load balancers. A rewriting middle layer will be required.

**Comment by kylegordon** [comment]
@RollingFred I just parked it sadly. I refuse (currently) to include a rewriting layer sidecar bodge with Grafana. Alternatives are welcome.

**Comment by RollingFred** [comment]
@kylegordon My goal was to use one ALB instead of multiple ELBs. I ended up creating a listener on port 3000 to which I attached my Grafana service, using '/' as a path. 
It works for us, as I have another NGINX reverse proxy in the front of that ALB anyway (for other purpose, like forwarding headers, security, etc...)
Grafana is configured with root_url="https://NGINX_URL:NGINX_PORT/grafana"
I end up with:
 NGINX_URL (/grafana) --> ALB_URL:3000 --> Grafana service
and same for other services.

**Comment by prandelicious** [comment]
I can't get the ALB to work with `/grafana` and just changing the `root_url` in the instance where Grafana lives. I had to add an nginx reverse proxy in front of Grafana for this to work.

**Comment by kattelus** [comment]
This is kind of sad that web context path can't be changed. Makes impossible to use aws elb with subpaths in front of grafana. If I am wrong please correct me?

**Comment by cortopy** [comment]
Same issues here using an ALB with grafana running in a kubernetes cluster. Not having grafana in subpath seems to work though

**Comment by sriraja53** [comment]
> I can't get the ALB to work with `/grafana` and just changing the `root_url` in the instance where Grafana lives. I had to add an nginx reverse proxy in front of Grafana for this to work.

Same issue but I am trying to use above approach.

**Comment by nomade1999** [comment]
Same here... we need to put it behind the same ALB with a Forward based on path.

I don't get the point of allowing to configure root_url if it does not work???

**Comment by ro6it** [comment]
Still having the same issue, when trying to configure a alb rule /grafana or /login it is not working at all, any way to use it behind the application load balancer without not tweaking much of the configuration.

**Comment by BryanGoble** [comment]
Not sure if this is still useful to anyone, but I managed to get this working and figured it would be worth sharing.

To set up an AWS ECS hosted Grafana instance behind an AWS ALB, setup the following:

ALB:
1. Listener rule 1
    a. Condition - Path Pattern is /grafana
    b. Action - Redirect to HTTPS://#{host}:443/#{path}/? - Status Code: HTTP_302
2. Listener rule 2
    a. Condition - Path Pattern is /grafana/*
    b. Action - Forward to target group <grafana-tg> - TG stickiness off

Grafana Task Definition: (Configure the following environment variables)
1. GF_SERVER_DOMAIN - if using ALB with existing domain and adding grafana as a child path, like I did
2. GF_SERVER_ENFORCE_DOMAIN - true
3. GF_SERVER_ROOT_URL - "%(protocol)s://%(domain)s/grafana/"
